<h2>Predicting the Estimated Time of Arrival for Bikes in Nairobi

BLOSSOM ML GROUP C - MARCH 2020</h2>

In [29]:
#importing libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [30]:
#import data to be used
train=pd.read_csv('Train.csv')
riders=pd.read_csv('Riders.csv')

In [31]:
#visualize dataframes
train

Order No       User Id Vehicle Type  Platform Type  \
0       Order_No_4211   User_Id_633         Bike              3   
1      Order_No_25375  User_Id_2285         Bike              3   
2       Order_No_1899   User_Id_265         Bike              3   
3       Order_No_9336  User_Id_1402         Bike              3   
4      Order_No_27883  User_Id_1737         Bike              1   
...               ...           ...          ...            ...   
21196   Order_No_8834  User_Id_2001         Bike              3   
21197  Order_No_22892  User_Id_1796         Bike              3   
21198   Order_No_2831  User_Id_2956         Bike              3   
21199   Order_No_6174  User_Id_2524         Bike              1   
21200   Order_No_9836   User_Id_718         Bike              3   

      Personal or Business  Placement - Day of Month  \
0                 Business                         9   
1                 Personal                        12   
2                 Business                        30   
3                 Business                        15   
4                 Personal                        13   
...                    ...                       ...   
21196             Personal                        20   
21197             Business                        13   
21198             Business                         7   
21199             Personal                         4   
21200             Business                        26   

       Placement - Weekday (Mo = 1) Placement - Time  \
0                                 5       9:35:46 AM   
1                                 5      11:16:16 AM   
2                                 2      12:39:25 PM   
3                                 5       9:25:34 AM   
4                                 1       9:55:18 AM   
...                             ...              ...   
21196                             3       3:54:38 PM   
21197                             6      10:13:34 AM   
21198                             4       5:06:16 PM   
21199                             3       9:31:39 AM   
21200                             2       2:19:47 PM   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  ...  \
0                                9                                5  ...   
1                               12                                5  ...   
2                               30                                2  ...   
3                               15                                5  ...   
4                               13                                1  ...   
...                            ...                              ...  ...   
21196                           20                                3  ...   
21197                           13                                6  ...   
21198                            7                                4  ...   
21199                            4                                3  ...   
21200                           26                                2  ...   

      Arrival at Destination - Time  Distance (KM)  Temperature  \
0                       10:39:55 AM              4         20.4   
1                       12:17:22 PM             16         26.4   
2                        1:00:38 PM              3          NaN   
3                       10:05:27 AM              9         19.2   
4                       10:25:37 AM              9         15.4   
...                             ...            ...          ...   
21196                    4:20:17 PM              3         28.6   
21197                   10:46:17 AM              7         26.0   
21198                    6:40:05 PM             20         29.2   
21199                   10:08:15 AM             13         15.0   
21200                    3:17:23 PM             12         30.9   

      Precipitation in millimeters  Pickup Lat  Pickup Long Destination Lat  \
0                              NaN   -1.317755    36.830370      

In [32]:
#visualize dataframes
riders

Rider Id  No_Of_Orders   Age  Average_Rating  No_of_Ratings
0    Rider_Id_396          2946  2298            14.0           1159
1    Rider_Id_479           360   951            13.5            176
2    Rider_Id_648          1746   821            14.3            466
3    Rider_Id_753           314   980            12.5             75
4    Rider_Id_335           536  1113            13.7            156
..            ...           ...   ...             ...            ...
955  Rider_Id_896           152    99            12.4             18
956  Rider_Id_149            69   101            10.2             10
957  Rider_Id_270           338    96            14.4             41
958  Rider_Id_201           159    96            15.0              9
959  Rider_Id_904           172   108            13.7             18

[960 rows x 5 columns]

In [33]:
#get summary of the train data
train.describe()

Platform Type  Placement - Day of Month  Placement - Weekday (Mo = 1)  \
count   21201.000000              21201.000000                  21201.000000   
mean        2.752182                 15.653696                      3.240083   
std         0.625178                  8.798916                      1.567295   
min         1.000000                  1.000000                      1.000000   
25%         3.000000                  8.000000                      2.000000   
50%         3.000000                 15.000000                      3.000000   
75%         3.000000                 23.000000                      5.000000   
max         4.000000                 31.000000                      7.000000   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  \
count                 21201.000000                     21201.000000   
mean                     15.653837                         3.240225   
std                       8.798886                         1.567228   
min                       1.000000                         1.000000   
25%                       8.000000                         2.000000   
50%                      15.000000                         3.000000   
75%                      23.000000                         5.000000   
max                      31.000000                         7.000000   

       Arrival at Pickup - Day of Month  Arrival at Pickup - Weekday (Mo = 1)  \
count                      21201.000000                          21201.000000   
mean                          15.653837                              3.240225   
std                            8.798886                              1.567228   
min                            1.000000                              1.000000   
25%                            8.000000                              2.000000   
50%                           15.000000                              3.000000   
75%                           23.000000                              5.000000   
max                           31.000000                              7.000000   

       Pickup - Day of Month  Pickup - Weekday (Mo = 1)  \
count           21201.000000               21201.000000   
mean               15.653837                   3.240225   
std                 8.798886                   1.567228   
min                 1.000000                   1.000000   
25%                 8.000000                   2.000000   
50%                15.000000                   3.000000   
75%                23.000000                   5.000000   
max                31.000000                   7.000000   

       Arrival at Destination - Day of Month  \
count                           21201.000000   
mean                               15.653837   
std                                 8.798886   
min                                 1.000000   
25%                                 8.000000   
50%                                15.000000   
75%                                23.000000   
max                                31.000000   

       Arrival at Destination - Weekday (Mo = 1)  Distance (KM)   Temperature  \
count                               21201.000000   21201.000000  16835.000000   
mean                                    3.240225       9.506533     23.258889   
std                                     1.567228       5.668963      3.615768   
min                                     1.000000       1.000000     11.200000   
25%                                     2.000000       5.000000     20.600000   
50%                                     3.000000       8.000000     23.500000   
75%                                     5.000000      13.000000     26.000000   
max                                     7.000000      49.000000     32.100000   

       Precipitation in millimeters    Pickup Lat   Pickup Long  \
count                    552.000000  21201.000000  21201.000000   
mean                       7.905797     -1.281470     36.811264   
std                 

In [34]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21201 entries, 0 to 21200
Data columns (total 29 columns):
Order No                                     21201 non-null object
User Id                                      21201 non-null object
Vehicle Type                                 21201 non-null object
Platform Type                                21201 non-null int64
Personal or Business                         21201 non-null object
Placement - Day of Month                     21201 non-null int64
Placement - Weekday (Mo = 1)                 21201 non-null int64
Placement - Time                             21201 non-null object
Confirmation - Day of Month                  21201 non-null int64
Confirmation - Weekday (Mo = 1)              21201 non-null int64
Confirmation - Time                          21201 non-null object
Arrival at Pickup - Day of Month             21201 non-null int64
Arrival at Pickup - Weekday (Mo = 1)         21201 non-null int64
Arrival at Pickup - Time   

<h4>For the temperature and precipitation in millimeters we realised that the data frame is missing some points</h4>

Temperature has a fill of 16835 / 21201 - including the heading<br>
Precipitation in millimeters has 552/21201 - including the heading

In [37]:
#Checking if the objects under Temperature and precipitation are still valuable to the analysis lets do a count of the uniqueness of the values in them using .nunique()
train.nunique()

Order No                                     21201
User Id                                       3186
Vehicle Type                                     1
Platform Type                                    4
Personal or Business                             2
Placement - Day of Month                        31
Placement - Weekday (Mo = 1)                     7
Placement - Time                             15686
Confirmation - Day of Month                     31
Confirmation - Weekday (Mo = 1)                  7
Confirmation - Time                          15742
Arrival at Pickup - Day of Month                31
Arrival at Pickup - Weekday (Mo = 1)             7
Arrival at Pickup - Time                     15767
Pickup - Day of Month                           31
Pickup - Weekday (Mo = 1)                        7
Pickup - Time                                15690
Arrival at Destination - Day of Month           31
Arrival at Destination - Weekday (Mo = 1)        7
Arrival at Destination - Time  

In [39]:
train['Temperature'].unique()

array([20.4, 26.4,  nan, 19.2, 15.4, 27.2, 20.3, 28.7, 24.6, 16.4, 26.1,
       20.9, 24.4, 19.4, 20.8, 23.3, 18.1, 22.5, 26.5, 27.3, 25.6, 20.5,
       18.8, 16.1, 25.3, 22.3, 26.3, 29.2, 20. , 18. , 21.6, 24.8, 23. ,
       28.1, 14.7, 25. , 26.7, 17.6, 22.7, 23.9, 24.7, 28.2, 19.1, 20.1,
       17.7, 17.3, 21.9, 24.5, 26.9, 18.4, 22. , 25.7, 23.7, 25.8, 16.2,
       31. , 26.2, 28.6, 19.3, 25.2, 24. , 25.5, 23.6, 30.1, 22.8, 27.7,
       21.3, 19.7, 14.5, 30.8, 23.8, 21.7, 24.2, 25.9, 21.2, 18.9, 17.9,
       18.7, 25.4, 22.9, 27.8, 22.6, 18.6, 26. , 21.1, 24.9, 18.5, 15.3,
       27.5, 22.4, 29.6, 19.9, 27.6, 19.8, 20.6, 16.8, 27.4, 24.1, 22.1,
       23.1, 27. , 29.8, 19.6, 21.8, 25.1, 16.9, 30. , 23.5, 23.2, 26.6,
       28.5, 17.8, 30.5, 15.1, 21. , 16. , 26.8, 24.3, 18.3, 27.1, 20.7,
       18.2, 23.4, 14.8, 15.8, 28.4, 21.4, 17.5, 29. , 19.5, 27.9, 22.2,
       28. , 20.2, 29.1, 17.1, 29.9, 15.6, 31.8, 28.3, 30.7, 17.2, 29.7,
       16.7, 19. , 29.3, 30.6, 29.4, 21.5, 16.6, 28

In [41]:
train['Precipitation in millimeters'].unique()

array([ nan,  3.9,  0.8,  0.4,  7. ,  1.1,  9.1,  1.9,  2.9,  4.9, 10.9,
        3. ,  0.2, 51.1, 15.9,  2. ,  0.9,  1. ,  3.1, 16.1,  5. ,  4. ,
        4.1, 11. ,  0.7, 51.9,  2.1, 83.1, 16. ,  0.1,  0.5,  0.6, 42.9,
       51. ,  8.9,  0.3, 11.1,  5.1, 52.1, 16.9, 83. , 50.9, 17. , 43. ,
       14. , 99.1, 17.1, 98.9, 82.9, 13.9, 43.1,  7.1, 14.9, 99. ,  9. ])

With this we see that the two columns have cells filled with nan
since the column contain other value it would still be valuable to the analysis  else we would have droped it using that train=train.drop(['Precipitation in millimeters','Temperature'],axis=1). Save new dataframes as train.to_csv("train_new.csv", index=false)

In [35]:
#get summary of the riders data
riders.describe()

No_Of_Orders          Age  Average_Rating  No_of_Ratings
count    960.000000   960.000000      960.000000     960.000000
mean    1864.851042  1200.234375       13.412604     427.983333
std     1880.337785   810.930171        2.675794     486.957931
min        2.000000    96.000000        0.000000       0.000000
25%      261.750000   478.250000       13.500000      30.000000
50%     1475.500000  1021.000000       14.000000     223.000000
75%     2847.250000  1891.500000       14.300000     678.750000
max     9756.000000  3764.000000       15.200000    2298.000000

In [36]:
riders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 960 entries, 0 to 959
Data columns (total 5 columns):
Rider Id          960 non-null object
No_Of_Orders      960 non-null int64
Age               960 non-null int64
Average_Rating    960 non-null float64
No_of_Ratings     960 non-null int64
dtypes: float64(1), int64(3), object(1)
memory usage: 37.6+ KB


In [38]:
#checking for how dynamic the values in the rider dataset is we used .nunique()
riders.nunique()

Rider Id          960
No_Of_Orders      811
Age               725
Average_Rating     45
No_of_Ratings     553
dtype: int64

In [43]:
#in merging the two dataframes into one using a comon field

sendy=train.merge(riders,how='left',on='Rider Id')
sendy

Order No       User Id Vehicle Type  Platform Type  \
0       Order_No_4211   User_Id_633         Bike              3   
1      Order_No_25375  User_Id_2285         Bike              3   
2       Order_No_1899   User_Id_265         Bike              3   
3       Order_No_9336  User_Id_1402         Bike              3   
4      Order_No_27883  User_Id_1737         Bike              1   
...               ...           ...          ...            ...   
21196   Order_No_8834  User_Id_2001         Bike              3   
21197  Order_No_22892  User_Id_1796         Bike              3   
21198   Order_No_2831  User_Id_2956         Bike              3   
21199   Order_No_6174  User_Id_2524         Bike              1   
21200   Order_No_9836   User_Id_718         Bike              3   

      Personal or Business  Placement - Day of Month  \
0                 Business                         9   
1                 Personal                        12   
2                 Business                        30   
3                 Business                        15   
4                 Personal                        13   
...                    ...                       ...   
21196             Personal                        20   
21197             Business                        13   
21198             Business                         7   
21199             Personal                         4   
21200             Business                        26   

       Placement - Weekday (Mo = 1) Placement - Time  \
0                                 5       9:35:46 AM   
1                                 5      11:16:16 AM   
2                                 2      12:39:25 PM   
3                                 5       9:25:34 AM   
4                                 1       9:55:18 AM   
...                             ...              ...   
21196                             3       3:54:38 PM   
21197                             6      10:13:34 AM   
21198                             4       5:06:16 PM   
21199                             3       9:31:39 AM   
21200                             2       2:19:47 PM   

       Confirmation - Day of Month  Confirmation - Weekday (Mo = 1)  ...  \
0                                9                                5  ...   
1                               12                                5  ...   
2                               30                                2  ...   
3                               15                                5  ...   
4                               13                                1  ...   
...                            ...                              ...  ...   
21196                           20                                3  ...   
21197                           13                                6  ...   
21198                            7                                4  ...   
21199                            4                                3  ...   
21200                           26                                2  ...   

      Pickup Lat  Pickup Long  Destination Lat Destination Long      Rider Id  \
0      -1.317755    36.830370        -1.300406        36.829741  Rider_Id_432   
1      -1.351453    36.899315        -1.295004        36.814358  Rider_Id_856   
2      -1.308284    36.843419        -1.300921        36.828195  Rider_Id_155   
3      -1.281301    36.832396        -1.257147        36.795063  Rider_Id_855   
4      -1.266597    36.792118        -1.295041        36.809817  Rider_Id_770   
...          ...          ...              ...              ...           ...   
21196  -1.258414    36.804800        -1.275285        36.802702  Rider_Id_953   
21197  -1.307143    36.825009        -1.331619        36.847976  Rider_Id_155   
21198  -1.286018    36.897534        -1.258414        36.804800  Rider_Id_697   
21199  -1.250030    36.874167        -1.279209        36.794872  Rider_Id_347   
21200  -1.255189    36.782203        -1.320157        36.8

from the table we can see the number of columns have increased to 33 due to the merge of the train and riders dataframes